scipy/special/xsf/legendre.h

```diff
@@ -9,38 +10,22 @@ template <typename T>
 struct legendre_p_initializer_n {
     T z;
 
-    void operator()(std::tuple<T (&)[2]> res) const { tuples::assign(res, {{1, z}}); }
-
-    void operator()(std::tuple<T (&)[2], T (&)[2]> res) const { tuples::assign(res, {{1, z}, {0, 1}}); }
-
-    void operator()(std::tuple<T (&)[2], T (&)[2], T (&)[2]> res) const {
-        tuples::assign(res, {{1, z}, {0, 1}, {0, 0}});
+    void operator()(T (&res)[2]) const {
+        res[0] = T(1);
+        res[1] = z;
     }
 };
 
 template <typename T>
 struct legendre_p_recurrence_n {
     T z;
 
-    void operator()(int n, std::tuple<T (&)[2]> res) const {
-        T fac0 = -T(n - 1) / T(n);
-        T fac1 = T(2 * n - 1) / T(n);
-
-        tuples::assign(res, {{fac0, fac1 * z}});
-    }
-
-    void operator()(int n, std::tuple<T (&)[2], T (&)[2]> res) const {
-        T fac0 = -T(n - 1) / T(n);
-        T fac1 = T(2 * n - 1) / T(n);
-
-        tuples::assign(res, {{fac0, fac1 * z}, {0, fac1}});
-    }
-
-    void operator()(int n, std::tuple<T (&)[2], T (&)[2], T (&)[2]> res) const {
+    void operator()(int n, T (&res)[2]) const {
         T fac0 = -T(n - 1) / T(n);
         T fac1 = T(2 * n - 1) / T(n);
```

Comment:
As we discussed offline, it's wasteful to perform these regular scalar operations on dual numbers. You'd suggested to define something like `real_type_t` for dual numbers, so we can get the correct type here. I checked and confirmed that doing so dramatically speeds things up.